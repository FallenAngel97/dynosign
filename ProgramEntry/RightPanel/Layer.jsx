import React from 'react';
import './Layer.scss';
import { connect } from 'react-redux';
import { changeActiveLayer, changeLayer, changeLayerVisibility } from '../actions'
import hiddenEye from './hidden_eye.svg';
import visibleEye from './visible_eye.svg';
import PropTypes from 'prop-types';

/**
 * Core module, which handles layer manipulation logic in the right panel
 * @module Layer
 */

export class Layer extends React.Component {

  state = {
    editLayer: false
  }
  
  _onKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.setState({ editLayer: false });
    }
  }
  showLayerOptions = (e) => {
    this.props.openContext(e);
  }
  UNSAFE_componentWillMount () {
    document.addEventListener('mousedown', this.handleClick, false);
  }
  DragOver = (e) => {
    this.props.layerDragOver(e);
  }
  ondragStart = (e) => {
    this.props.layerDragStart(e);
  }
  dragEnd = (e) => {
    this.props.layerDragEnd(e);
  }
  ondrop = (e) => {
    this.props.layerDrop(e);
  }
  dragLeave = (e) => {
    this.props.layerDragLeave(e);
  }
  componentWillUnmount () {
    document.removeEventListener('mousedown', this.handleClick, false);
  }
  changeVisibility = () => {
    const { layer } = this.props;
    this.props.changeLayerVisibility(!layer.hidden, this.props.layerId);
  }
  /**
   * Hides the text rename box, if clicked outside. If clicked on the layer single time,
   * changes the active layer to the clicked one
   * @param {MouseEvent} ev - neccessary to check, what is the clicked target
   */
  handleClick = (ev) => {
    if (!this.node.contains(ev.target)) {
      this.setState({ editLayer: false })
    } else if (ev.target.className.indexOf('singleLayer') > -1) {
      this.props._changeActiveLayer(this.props.layer, this.props.layerId);
    }
  }
  /**
   * Opens the rename text box
   */
  renameLayer = () => {
    this.setState({ editLayer: true });
  }
  layerNameChange = (ev) => {
    let { layer } = this.props.changeActiveLayer;
    layer.name = ev.target.value;
    this.props.changeLayer(layer, this.props.changeActiveLayer.layerId)
  }
  render () {
    const { editLayer } = this.state;
    const { changeActiveLayer, layerId, layer } = this.props;

    return (
      <div onDragStart={this.ondragStart} draggable={!editLayer}
        ref={node => { this.node = node }}
        onDoubleClick={this.renameLayer}
        onDrop={this.ondrop}
        onKeyPress={this._onKeyPress}
        onContextMenu={this.showLayerOptions}
        onDragLeave={this.dragLeave}
        onDragOver={this.DragOver}
        onDragEnd={this.dragEnd}
        className={'singleLayer' + ((changeActiveLayer.layerNumber === layerId) ? ' activeLayer' : '')}>
        <img onClick={this.changeVisibility} src={layer.hidden ? hiddenEye : visibleEye}/>
        {editLayer ? <input onChange={this.layerNameChange} type='text' className='singleLayerEdit' value={layer.name} /> : layer.name}
      </div>
    )
  }
}

Layer.propTypes = {
  layer: PropTypes.object,
  changeLayerVisibility: PropTypes.func,
  layerId: PropTypes.number,
  _changeActiveLayer: PropTypes.func,
  changeActiveLayer: PropTypes.object,
  changeLayer: PropTypes.func,
  layerDragStart: PropTypes.func,
  layerDragEnd: PropTypes.func,
  layerDrop: PropTypes.func,
  layerDragOver: PropTypes.func,
  layerDragLeave: PropTypes.func,
  openContext: PropTypes.func
}

const mapStateToProps = state => state;

export const mapDispatchToProps = dispatch => {
  return {
    _changeActiveLayer: (layer, layerNumber) => dispatch(changeActiveLayer(layer, layerNumber)),
    changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
    changeLayerVisibility: (hidden, layerId) => dispatch(changeLayerVisibility(hidden, layerId))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Layer);
