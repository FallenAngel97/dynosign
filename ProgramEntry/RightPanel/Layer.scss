.activeLayer {
    background: rgba(139, 146, 99, 0.2);
    input {
        width: 159px;
    }
}
.singleLayer{
    transition: .2s ease-in-out padding;
    img {
        height: 11px;
    }
}
.drag_under {
    padding-top: 8px;
    padding-bottom: 8px;
    color: red;
}