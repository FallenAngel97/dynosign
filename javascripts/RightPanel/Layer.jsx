import React from "react";
import "./Layer.scss";
import {connect} from "react-redux";
import {changeActiveLayer, changeLayer, changeLayerVisibility} from "../actions"
import hidden_eye from "./hidden_eye.svg";
import visible_eye from "./visible_eye.svg";

export class Layer extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            editLayer: false
        }
        this.renameLayer = this.renameLayer.bind(this);
        this.handleClick = this.handleClick.bind(this);
        this.layerNameChange = this.layerNameChange.bind(this);
        this.changeVisibility = this.changeVisibility.bind(this);
    }
    componentWillMount() {
        document.addEventListener('mousedown', this.handleClick, false);
    }
    componentWillUnmount() {
        document.removeEventListener('mousedown', this.handleClick, false);
    }
    changeVisibility() {
        let layer = this.props.layer;
        this.props.changeLayerVisibility(!layer.hidden, this.props.layerId);
    }
    handleClick(ev) {
        if(!this.node.contains(ev.target) ){
            this.setState({editLayer: false})
        } else if(ev.target.className.indexOf("singleLayer")>-1) {
            this.props._changeActiveLayer(this.props.layer, this.props.layerId);
        }
    }
    renameLayer() {
        this.setState({editLayer: true});
    }
    layerNameChange(ev) {
        let layer = this.props.changeActiveLayer.layer;
        layer.name = ev.target.value;
        this.props.changeLayer(layer, this.props.changeActiveLayer.layerId)
    }
    render() {
        return (
            <div ref={node=>this.node=node} onDoubleClick={this.renameLayer}
                className={"singleLayer"+((this.props.changeActiveLayer.layerNumber === this.props.layerId) ? ' activeLayer':'')}>
                <img onClick={this.changeVisibility} src={this.props.layer.hidden ? hidden_eye : visible_eye}/>
                {this.state.editLayer ? <input onChange={this.layerNameChange} type='text' className='singleLayerEdit' value={this.props.layer.name} /> : this.props.layer.name}
            </div>
        )
    }
}

const mapStateToProps = state => state;

const mapDispatchToProps = dispatch => {
    return {
        _changeActiveLayer: (layer,layerNumber) => dispatch(changeActiveLayer(layer,layerNumber)),
        changeLayer: (layer, layerId) => dispatch(changeLayer(layer, layerId)),
        changeLayerVisibility: (hidden, layerId) => dispatch(changeLayerVisibility(hidden, layerId))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Layer);