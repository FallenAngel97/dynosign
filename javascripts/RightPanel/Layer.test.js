import React from "react";
import { Layer, mapDispatchToProps } from "./Layer";
import { configure, shallow, render} from "enzyme"
import Adapter from 'enzyme-adapter-react-16';

const defaultLayer = {
    opacity: 100,
    name: 'New layer',
    hidden: false,
    linesArray: []
}

configure({ adapter: new Adapter() })

describe("Layer component", () => {
    test("Layer should render properly", () => {
        const layer = shallow(<Layer changeActiveLayer={{layerNumber: 0}} layer={{name: 'Layer 0'}}  />);
        expect(layer).toMatchSnapshot();
    });
    test("Layer should allow editing", () => {
        const wrapper = shallow(<Layer changeActiveLayer={{layerNumber: 0}} layer={{name: 'Layer 0'}} />);
        wrapper.simulate('dblclick');
        expect(wrapper.state().editLayer).toEqual(true);
    });
    test("Layer component can trigger 'change active layer'", () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch)._changeActiveLayer(defaultLayer, 1);
        expect(dispatch.mock.calls[0][0]).toEqual(
            { type: 'CHANGE_ACTIVE_LAYER', layer: defaultLayer, layerNumber: 1}
        );
    });
    test("Layer component can trigger 'change layer'", () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).changeLayer(defaultLayer, 1);
        expect(dispatch.mock.calls[0][0]).toEqual(
            { type: 'CHANGE_LAYER', layer: defaultLayer, layerId: 1}
        );
    });
    test("Layer component can trigger 'change layer visibility'", () => {
        const dispatch = jest.fn();
        mapDispatchToProps(dispatch).changeLayerVisibility(true, 1);
        expect(dispatch.mock.calls[0][0]).toEqual(
            {type: 'CHANGE_LAYER_VISIBILITY', hidden: true, layerNumber: 1}
        );
    });
    test("Layer should allow renaming", () => {
        const wrapper = shallow(<Layer changeActiveLayer={{layerNumber: 0}} layer={{name: 'Layer 0'}}  />);
        wrapper.instance().renameLayer();
        expect(wrapper.state().editLayer).toEqual(true);
    });
    test("Layer should handle click", () => {
        const wrapper = shallow(<Layer changeActiveLayer={{layerNumber: 0}} layer={{name: 'Layer 0'}}  />);
        wrapper.simulate('click');
        expect(wrapper.state().editLayer).toEqual(false);
    });
});