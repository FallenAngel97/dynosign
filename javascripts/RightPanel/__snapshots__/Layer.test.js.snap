// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer component Layer should render properly 1`] = `ShallowWrapper {}`;
